KAFKA-7210:  Add a system test to verify the log compaction

omkreddy opened the pull request:
- Update TestLogCleaning tool to use Java consumer and rename as LogCompactionTester
- Enable log compaction in System tests
- Remove configs with values same as server defaults from "kafka.properties" template file.
- Update the kafka.py logic to handle the duplicates between kafka.properties and server_prop_overides.

omkreddy commented:
@ijuma  Pls take a look. I have  added a system test to verify compaction using this tool.  
It will be good if we can merge this along with old consumer changes.

omkreddy commented:
retest this please

ijuma commented on tests/kafkatest/services/kafka/templates/kafka.properties:
```diff
@@ -29,8 +29,6 @@ socket.request.max.bytes=104857600
 num.recovery.threads.per.data.dir=1
 log.retention.hours=168
-log.segment.bytes=1073741824
-log.cleaner.enable=false
```
Why do we have to do this?

omkreddy replied:
"kafka.properties" template file contains default config values to populate server.properties.

Now log cleaner is enabled by default. So I have removed this property to enable it by default is tests.  

"log.segment.bytes"  property configured with  sever default value and not updated by any of the test. 

 For now I reverted this change.  if required we can do this cleanup later.

Also updated the logic in kafka.py to cleanup the duplicates between kafka.properties and server_prop_overides. Updated the logic to load the kafka.properties first and then override any server_prop_overides configs.

ijuma replied:
I think it's good to enable the cleaner in the system tests. And if the other property is just the default, I agree we should just remove it from here.

ijuma reviewed (commented):
Thanks @omkreddy, this is very useful. @cmccabe, do you think you could review this?

cmccabe commented on tests/kafkatest/services/test_log_compaction_tool.py:
```diff
@@ -0,0 +1,74 @@
+
+    logs = {
+        "logcompaction_tool_stdout": {
+            "path": "/mnt/logcompaction_tool_stdout.log",
```
We should a subdirectory of /mnt, not directly inside /mnt

omkreddy replied:
Updated the code

cmccabe commented on tests/kafkatest/services/test_log_compaction_tool.py:
```diff
@@ -0,0 +1,74 @@
+        cmd += " %s" % self.java_class_name()
+        cmd += " --broker %s --messages 1000000 --sleep 20 --duplicates 10 --percent-deletes 10" % (self.kafka.bootstrap_servers(self.security_protocol))
+
+        cmd += " 2>> /mnt/logcompaction_tool_stdout.log | tee -a /mnt/logcompaction_tool_stdout.log &"
```
Can we just use &>> ?

omkreddy replied:
with "&>>" I am unable to run the command as background job. I will try again.

cmccabe replied:
It's fine to leave it as-is if bash is being fiddly.

cmccabe commented on tests/kafkatest/services/test_log_compaction_tool.py:
```diff
@@ -0,0 +1,74 @@
+from kafkatest.directory_layout.kafka_path import KafkaPathResolverMixin
+from kafkatest.services.security.security_config import SecurityConfig
+
+class TestLogCompactionTool(KafkaPathResolverMixin, BackgroundThreadService):
```
This should be named LogCompactionTool.  It is not a test (it is a service used by tests)

omkreddy commented:
@cmccabe  Thanks for the review.  Addressed the review comments.

Also updated the logic in kafka.py to cleanup the duplicates between kafka.properties and server_prop_overides. We load the kafka.properties first and then override any server_prop_overides configs.

Pls take a look.

ijuma commented on core/src/main/scala/kafka/tools/TestLogCompactionTool.scala:
```diff
@@ -46,64 +51,60 @@ import scala.collection.JavaConverters._
  * print an error and exit with exit code 1, otherwise we print the size reduction and exit with exit code 0.
  */
-object TestLogCleaning {
+object TestLogCompactionTool {
```
This should remain in the test directory right? It's similar to `MiniKdc`.

omkreddy replied:
It is similar to ReplicaVerificationTool. So i moved to src directory.  This will be useful, if we want to test manually with binary distribution.

Also, adding testlibs to minikdc service classpath handled separately. not sure we want do same here.
https://github.com/apache/kafka/blob/trunk/tests/kafkatest/services/security/minikdc.py#L110MiniKdc

ijuma replied:
A KIP is needed for adding new user visible tools. Personally, I don't see much value in testing the binary distribution manually.

omkreddy replied:
Ok..  moved tool class to test directory and retained the original name.

omkreddy committed:
MINOR: Update TestLogCleaning tool to use java consumer

omkreddy commented:
@ijuma @cmccabe  rebased the PR.  Verified most of the system tests on my local machine. 
Is it possible to start system test run for this branch?

omkreddy committed:
Use adminclient to create compacted topics

ijuma commented:
@cmccabe, can you please review?

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+      .describedAs("count")
+      .ofType(classOf[java.lang.Integer])
+      .defaultsTo(5)
+    val brokerOpt = parser.accepts("broker", "Url to connect to.")
```
The should be `--bootstrap-server` to match how our tools work.  (Some of them use nonstandard synonyms currently, but we want to converge on --bootstrap-server)

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+      .describedAs("percent")
+      .ofType(classOf[java.lang.Integer])
+      .defaultsTo(0)
+    val sleepSecsOpt = parser.accepts("sleep", "Time to sleep between production and consumption.")
```
Can we specify in the help text that this is the time in milliseconds?

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+    val topicCount = options.valueOf(topicsOpt).intValue
+    val sleepSecs = options.valueOf(sleepSecsOpt).intValue
+
+    val testId = new Random().nextInt(Int.MaxValue)
```
Can we use nextLong here to be more certain we won't get a collision?

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+  def createTopics(brokerUrl: String, topics: Seq[String]): Unit = {
+    val adminConfig = new Properties
+    adminConfig.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, brokerUrl)
+    val adminClient = admin.AdminClient.create(adminConfig)
```
We should have a try/catch block here to prevent the AdminClient object from leaking.

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+    producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+    producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+    producerProps.setProperty(ProducerConfig.COMPRESSION_TYPE_CONFIG, compressionType)
+    val producer = new KafkaProducer[Array[Byte], Array[Byte]](producerProps)
```
let's have a try/catch block here as well to prevent a leak.

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+    val keyCount = (messages / dups).toInt
+    val producedFile = File.createTempFile("kafka-log-cleaner-produced-", ".txt")
+    println("Logging produce requests to " + producedFile.getAbsolutePath)
+    val produceFileWriter : OutputStreamWriter = new OutputStreamWriter(new FileOutputStream(producedFile), UTF_8)
```
Now that we're using Java8, you can just use `Files.newBufferedWriter` here

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+    for (i <- 0L until (messages * topics.length)) {
+      val topic = topics((i % topics.length).toInt)
+      val key = rand.nextInt(keyCount)
+      val delete = i % 100 < percentDeletes
```
can you add parentheses to make this expression clearer

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+          new ProducerRecord[Array[Byte], Array[Byte]](topic, key.toString.getBytes(UTF_8), null)
+        else
+          new ProducerRecord[Array[Byte], Array[Byte]](topic, key.toString.getBytes(UTF_8), i.toString.getBytes(UTF_8))
+      producer.send(msg)
```
I think the tool should fail if the produce operation fails?  We should check the result of the future, or callback, etc.

omkreddy replied:
test fails at data validation phase if producer failed to send data. I am going leave this as it is.

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+    val consumed = valuesIterator(consumedReader)
+
+    val producedDedupedFile = new File(producedDataFile.getAbsolutePath + ".deduped")
+    val produceWriter : OutputStreamWriter = new OutputStreamWriter(new FileOutputStream(producedDedupedFile), UTF_8)
```
Can use Files.newBufferedWriter here

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+    val producedDeduped = new BufferedWriter(produceWriter, 1024 * 1024)
+
+    val consumedDedupedFile = new File(consumedDataFile.getAbsolutePath + ".deduped")
+    val consumeWriter : OutputStreamWriter = new OutputStreamWriter(new FileOutputStream(consumedDedupedFile), UTF_8)
```
Can use Files.newBufferedWriter here

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+    val topics = (0 until topicCount).map("log-cleaner-test-" + testId + "-" + _).toArray
+    createTopics(brokerUrl, topics.toSeq)
+
+    println("Producing %d messages..to topics %s".format(messages, topics.mkString(",")))
```
It's better to use https://docs.scala-lang.org/overviews/core/string-interpolation.html

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+    }
+    producedDeduped.close()
+    consumedDeduped.close()
+    println("Validated " + total + " values, " + mismatched + " mismatches.")
```
Better to use https://docs.scala-lang.org/overviews/core/string-interpolation.html

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+  }
+
+  def peekLine(reader: BufferedReader) = {
+    reader.mark(4096)
```
Are we assuming that lines can't go over 4096 bytes?  This should be a constant somewhere if so, and commented

omkreddy replied:
yes, added constant with a comment.

cmccabe commented on core/src/test/scala/kafka/tools/TestLogCleaning.scala:
```diff
@@ -0,0 +1,361 @@
+    producedDeduped.close()
+    consumedDeduped.close()
+    println("Validated " + total + " values, " + mismatched + " mismatches.")
+    require(!produced.hasNext, "Additional values produced not found in consumer log.")
```
I don't think require is quite the right thing here, since that's similar to "assert" -- supposedly for debugging.  We probably should just have our own function that exits with a nonzero return code if the test fails.

omkreddy replied:
added a function to verify condition and exits with a nonzero return code if condition  fails.

cmccabe commented on tests/kafkatest/services/log_compaction_tool.py:
```diff
@@ -0,0 +1,87 @@
+        self.logger.info("TestLogCleaning tool %d command: %s" % (idx, cmd))
+        self.security_config.setup_node(node)
+        for line in node.account.ssh_capture(cmd):
+            self.logger.info("Checking line:{}".format(line))
```
This should be a debug log message, right?

cmccabe commented on tests/kafkatest/services/log_compaction_tool.py:
```diff
@@ -0,0 +1,87 @@
+        for line in node.account.ssh_capture(cmd):
+            self.logger.info("Checking line:{}".format(line))
+
+            if "Data verification is completed" in line:
```
We should have a constant for this.  And probably check that the line starts with it?

omkreddy replied:
done

cmccabe commented on tests/kafkatest/tests/tools/log_compaction_test.py:
```diff
@@ -0,0 +1,69 @@
+            security_protocol=security_protocol,
+            interbroker_security_protocol=interbroker_security_protocol,
+            server_prop_overides=[
+                [config_property.LOG_SEGMENT_BYTES, "1073000"],
```
How was this number chosen?  Should there be a constant?

omkreddy replied:
Configured smaller segment size  (1MB, default is 1GB) to create more segments  quickly for compaction. moved to constant.

cmccabe commented on tests/kafkatest/tests/tools/log_compaction_test.py:
```diff
@@ -0,0 +1,69 @@
+    @cluster(num_nodes=4)
+    def test_log_compaction(self, security_protocol='PLAINTEXT'):
+        """
+        Tests Log Compaction
```
This comment doesn't seem to add any information beyond repeating the function name.  Let's add something here or delete it.

omkreddy replied:
deleted the comment.

cmccabe commented:
Thanks, @omkreddy .  Posted some comments

omkreddy committed:
Address colin's comments

omkreddy commented:
@cmccabe Thanks for the detailed review. Updated the PR. Ran few systems tests to verify the changes. Please take a look.

omkreddy commented:
retest this please

cmccabe commented:
Thanks @omkreddy, it looks really good.  I ran the test locally using `ducker-ak` and it passed.

One last thing.  There still seems to be a bit of name inconsistency.  The Java class is called `kafka.tools.TestLogCleaning`, but the python class is named `LogCompactionTool`.  A log message refers to `TestLogCleaning tool`.

The Java class isn't really a "tool" since tools are something we expect users to run outside of a testing context.  It's also not really a "test" in the sense of being something you directly run.

Maybe it makes sense to name the Java class `LogCompactionTester`.  Then the Python service class would be `LogCompactionTester`.  The Python test that uses the service class could be `LogCompactionTest`.  What do you think?

omkreddy committed:
Update file names

omkreddy commented:
@cmccabe makes sense. Updated the class names and verified the system test. Pls take a look.

cmccabe commented:
LGTM.  Thanks, @omkreddy

omkreddy commented:
@ijuma Can you please take a look?

ijuma commented on build.gradle:
```diff
@@ -768,6 +768,7 @@ project(':core') {
       include('*.jar')
     }
     into "$buildDir/dependant-testlibs"
+    from (project(':clients').testJar ) { "$buildDir/dependant-testlibs" }
```
Can you please remind me why we need to do this again?

omkreddy replied:
I am using TestUtils.waitUntilTrue() method in the test tool, which intern depends on clients/test/TestUtils.java. I added this line include client test library to system test dependent libs,

ijuma replied:
Why doesn't the following do it?

```groovy
from (configurations.testRuntime) {
      include('*.jar')
}
```

It seems like it should include every runtime dependency jar?

omkreddy replied:
current "configurations.testRuntime" does not include dependent project (clients) test jar/dependencies.

Updated build.gradle based on below link, to include dependent project test dependencies.

https://softnoise.wordpress.com/2014/09/07/gradle-sub-project-test-dependencies-in-multi-project-builds/

We can decide whether to have this commit (or) previous one line change.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+    val options = parser.parse(args: _*)
+
+    if (args.length == 0)
+      CommandLineUtils.printUsageAndDie(parser, "An integration test for log cleaning.")
```
I'm confused about this message. Shouldn't we explain that arguments are expected or something like that?

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+      .describedAs("ms")
+      .ofType(classOf[java.lang.Integer])
+      .defaultsTo(0)
+    val dumpOpt = parser.accepts("dump", "Dump the message contents of a topic partition that contains test data from this test to standard out.")
```
Is this used by anything?

omkreddy replied:
this is part of legacy tool which can be used to dump the log contents (Similar dump log tool). We can remove If we think it is not required.

ijuma replied:
Yeah, if we don't have an immediate use, I'd remove it.

omkreddy replied:
removed the dumpOpt option

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+    val producedLines = lineCount(producedDataFile)
+    val consumedLines = lineCount(consumedDataFile)
+    val reduction = 1.0 - consumedLines.toDouble / producedLines.toDouble
+    println("%d rows of data produced, %d rows of data consumed (%.1f%% reduction).".format(producedLines, consumedLines, 100 * reduction))
```
Maybe we can use the `f` interpolator here https://docs.scala-lang.org/overviews/core/string-interpolation.html

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+      TestUtils.waitUntilTrue(() => {
+        val newTopics = adminClient.listTopics.names.get()
+        topics.forall(topicName => newTopics.contains(topicName))
+      }, "timed out waiting for topics")
```
It would be helpful to know which topics have not been created.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+    }
+  }
+
+  def lineCount(file: File): Int = io.Source.fromFile(file).getLines.size
```
We're leaking a file handle here, I suggest `Files.readAllLines`.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+  }
+
+  def externalSort(file: File): BufferedReader = {
+    val builder = new ProcessBuilder("sort", "--key=1,2", "--stable", "--buffer-size=20%", "--temporary-directory=" + System.getProperty("java.io.tmpdir"), file.getAbsolutePath)
```
Might be worth using `Files.createTempDirectory` instead of `System.getProperty("java.io.tmpdir")`

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+    producerProps.setProperty(ProducerConfig.MAX_BLOCK_MS_CONFIG, Long.MaxValue.toString)
+    producerProps.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerUrl)
+    producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+    producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
```
Nit: I suggest passing these as constructor parameters.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+      val keyCount = (messages / dups).toInt
+      val producedFilePath = Files.createTempFile("kafka-log-cleaner-produced-", ".txt")
+      println(s"Logging produce requests to $producedFilePath")
+      val producedWriter : BufferedWriter = Files.newBufferedWriter(producedFilePath, UTF_8)
```
Nit: there should be no space before `:`.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+          if (delete)
+            new ProducerRecord[Array[Byte], Array[Byte]](topic, key.toString.getBytes(UTF_8), null)
+          else
+            new ProducerRecord[Array[Byte], Array[Byte]](topic, key.toString.getBytes(UTF_8), i.toString.getBytes(UTF_8))
```
The type arguments aren't needed here.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+    consumerProps.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "log-cleaner-test-" + new Random().nextInt(Int.MaxValue))
+    consumerProps.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerUrl)
+    consumerProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringDeserializer")
+    consumerProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringDeserializer")
```
Nit: pass as constructor parameters. More concise and allows the type arguments to be inferred.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+          for (record <- consumerRecords.asScala) {
+            val delete = record.value() == null
+            val value = if (delete) -1L else record.value().toLong
+            consumedWriter.write(TestRecord(record.topic(), record.key.toInt, value, delete).toString)
```
Nit: a few unnecessary `()`

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+    val bytes = new Array[Byte](buffer.remaining)
+    buffer.get(bytes)
+    new String(bytes, UTF_8)
+  }
```
Can we just use `Utils.utf8`?

ijuma replied:
There's a version that takes a `ByteBuffer`, so you can simplify more.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,373 @@
+
+case class TestRecord(topic: String, key: Int, value: Long, delete: Boolean) {
+  def this(pieces: Array[String]) = this(pieces(0), pieces(1).toInt, pieces(2).toLong, pieces(3) == "d")
+  def this(line: String) = this(line.split("\t"))
```
These should probably be `parse` methods in the companion object (if they are used).

omkreddy replied:
added companion object. not sure, if this is what you're suggesting. Please take a look.

ijuma commented on tests/kafkatest/services/kafka/templates/kafka.properties:
```diff
@@ -20,17 +20,7 @@ advertised.host.name={{ node.account.hostname }}
-num.network.threads=3
-num.io.threads=8
-socket.send.buffer.bytes=102400
 socket.receive.buffer.bytes=65536
```
What's the default for this?

omkreddy replied:
default is : 102400, not sure if we can remove this line

ijuma replied:
I suggest we remove it.

ijuma commented on tests/kafkatest/tests/tools/log_compaction_test.py:
```diff
@@ -0,0 +1,66 @@
+        self.start_test_log_compaction_tool(security_protocol)
+
+        # Verify that compacted data verification completed in LogCompactionTester
+        wait_until(lambda: self.compaction_verifier.is_done, timeout_sec=120, err_msg="Timed out waiting to complete compaction")
```
How long does this test usually take? Are we sure that 2 minutes is enough?

omkreddy replied:
For the configured system test inputs (1million small messages) it will take 60seconds to complete.

ijuma replied:
OK, let's make it 180 seconds then.

ijuma reviewed (commented):
Thanks for the PR, it will be great to have this. I have a few comments.

omkreddy committed:
Address latest review comments

omkreddy commented:
@ijuma  Thanks for the review.  Updated the review . Pls take one more look.

omkreddy commented:
retest this please

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -153,12 +154,13 @@ object LogCompactionTester {
-        topics.forall(topicName => newTopics.contains(topicName))
-      }, "timed out waiting for topics")
+        allTopics = adminClient.listTopics.names.get.asScala.toSeq
+        topics.forall(topicName => allTopics.contains(topicName))
```
Seems like you can do the `filter` here instead of `forall` and then the message creation below becomes simpler.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -284,20 +286,18 @@ object LogCompactionTester {
-    producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
     producerProps.setProperty(ProducerConfig.COMPRESSION_TYPE_CONFIG, compressionType)
-    val producer = new KafkaProducer[Array[Byte], Array[Byte]](producerProps)
+    val producer = new KafkaProducer[Array[Byte], Array[Byte]](producerProps, new ByteArraySerializer, new ByteArraySerializer)
```
Nit: you can remove `[Array[Byte], Array[Byte]]`

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -322,13 +322,11 @@ object LogCompactionTester {
-    consumerProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringDeserializer")
     consumerProps.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
-    new KafkaConsumer[String, String](consumerProps)
+    new KafkaConsumer[String, String](consumerProps, new StringDeserializer, new StringDeserializer)
```
Nit: you can remove `[String, String]`.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -360,14 +358,17 @@ object LogCompactionTester {
 }
+
+object TestRecord {
+  def apply(line: String): TestRecord = new TestRecord(line.split("\t"))
```
The idea was to call this method `parse` to make it clear that it's parsing the line and can fail.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -360,14 +358,17 @@ object LogCompactionTester {
 }
 
 case class TestRecord(topic: String, key: Int, value: Long, delete: Boolean) {
   def this(pieces: Array[String]) = this(pieces(0), pieces(1).toInt, pieces(2).toLong, pieces(3) == "d")
```
Is this used outside of the `TestRecord.apply` method? If so, I'd move it to the companion object too and give it another name (maybe `parseComponents`). Otherwise, I'd just inline it.

omkreddy replied:
Updated the code. Pls take a look.

ijuma commented on core/src/test/scala/kafka/tools/LogCompactionTester.scala:
```diff
@@ -0,0 +1,374 @@
+    val options = parser.parse(args: _*)
+
+    if (args.length == 0)
+      CommandLineUtils.printUsageAndDie(parser, "A tool to test log compaction. Valid options are: ")
```
I think I'd say something about the fact that no arguments were passed, but some are expected.

omkreddy replied:
This line used for printing usage. next line checks for required arguments. Same pattern we are using in other tools.

ijuma reviewed (commented):
Thanks for the updates.

omkreddy committed:
Address review comments

omkreddy committed:
Add client test jar/dependencies to core/ configurations

omkreddy committed:
Revert "Add client test jar/dependencies to core/ configurations"

This reverts commit 02e8edf092bf2629f97b6332362d9d912a322ffc.

omkreddy committed:
Remove apply method

ijuma commented on build.gradle:
```diff
@@ -872,6 +873,16 @@ project(':clients') {
+  }
+
+  artifacts {
+    testOutput jarTest
```
Interesting. Does this create a new artifact? I think maybe we should go back to your previous solution that is limited to `copyDependantTestLibs` with a comment explaining why it's needed. We can consider separately whether we should move away from `testCompile project(':clients').sourceSets.test.output` in favour of the more complete solution across the board (we use test dependencies in multiple modules).

ijuma replied:
There's also the option of just using `https://plugins.gradle.org/plugin/com.github.hauner.jarTest/1.0`, but it's best to discuss that in a separate PR.

omkreddy replied:
makes sense. reverted the last commit. Will handle this in separate PR.
For now, added a comment to explain the need for previous solution.

omkreddy committed:
Add a comment to build.gradle

ijuma reviewed (approved):
Thanks for your patience, LGTM.
